src/stream/src/executor/sink.rs

```diff
@@ -133,22 +173,68 @@ impl SinkExecutor {
                         // At this point (instead of the point above when we receive the upstream
                         // data chunk), we make sure that we do have data to send out, and we can
                         // thus mark the txn as started.
-                        if !in_transaction {
-                            sink.begin_epoch(epoch).await?;
-                            in_transaction = true;
-                        }
-
-                        if let Err(e) = sink.write_batch(chunk.clone()).await {
-                            sink.abort().await?;
-                            return Err(e.into());
-                        }
-                        empty_checkpoint_flag = false;
+                        log_writer.write_chunk(chunk.clone()).await?;
 
                         yield Message::Chunk(chunk);
                     }
                 }
                 Message::Barrier(barrier) => {
-                    if barrier.checkpoint {
+                    log_writer
+                        .flush_current_epoch(barrier.epoch.curr, barrier.checkpoint)
+                        .await?;
```

Comment:
Maybe we should add some comments here to clarify the behavior: this call to `flush_current_epoch` will wait for the log reader to finish consuming all the data in current checkpoint epoch.

Reply:
Actually this is not expected for log store interface. This is only the behavior of the current in memory log store. With kv log store implemented, we can flush the current epoch without waiting for the log reader to finish consuming the data so that the sink can be decoupled with the upstream. But in either way, the current implementation of sink executor is the same for either implementation.